glusterd-volgen: Add functionality to accept any custom xlator

bonfffire opened the pull request:
Add new function which allow users to insert any custom xlators.
It makes to provide a way to add any processing into file operations.

This PR is fixed #1943

Users can deploy the plugin(xlator shared object) and integrate it to glusterfsd.

If users want to enable a custom xlator, do the follows:

1. put xlator object(.so file) into "XLATOR_DIR/user/"
2. set the option user.xlator.\<xlator\> to the existing xlator-name to specify of the position in graph
3. restart gluster volume

Options for custom xlator are able to set in "user.xlator.\<xlator\>.\<optkey\>".

The server_graph_builder() routine can insert the debug/* xlator into any layer in graph.
This patch is also based on that mechanism and is very similar.

In loop of the server_graph_builder(), check that "user.xlator.\<xlator\>" option value matches the xlator name(defined for debug/* xlators).
If custom xlator is founds that shoud be inserted to graph, allocates a new xlator_t for custom xlator and inserts into graph.

At this time, if the custom xlator options are found as well, and puts value to the xlator_t by calling xlator_set_option().

bonfffire committed:
glusterd-volgen: Add functionality to accept any custom xlator

Add new function which allow users to insert any custom xlators.
It makes to provide a way to add any processing into file operations.

This PR is fixed #1943

Users can deploy the plugin(xlator shared object) and integrate it to glusterfsd.

If users want to enable a custom xlator, do the follows:

1. put xlator object(.so file) into "XLATOR_DIR/user/"
2. set the option user.xlator.<xlator> to the existing xlator-name to specify of the position in graph
3. restart gluster volume

Options for custom xlator are able to set in "user.xlator.<xlator>.<optkey>".

The server_graph_builder() routine can insert the debug/* xlator into any layer in graph.
This patch is also based on that mechanism and is very similar.

In loop of the server_graph_builder(), check that "user.xlator.<xlator>" option value matches the xlator name(defined for debug/* xlators).
If custom xlator is founds that shoud be inserted to graph, allocates a new xlator_t for custom xlator and inserts into graph.

At this time, if the custom xlator options are found as well, and puts value to the xlator_t by calling xlator_set_option().

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
Can one of the admins verify this patch?

mykaul commented on xlators/mgmt/glusterd/src/glusterd-volgen.c:
```diff
@@ -2755,6 +2755,114 @@ check_and_add_debug_xl(volgen_graph_t *graph, dict_t *set_dict, char *volname,
     return ret;
 }
 
+gf_boolean_t
```
```suggestion
static gf_boolean_t
```

mykaul commented on xlators/mgmt/glusterd/src/glusterd-volgen.c:
```diff
@@ -2755,6 +2755,114 @@ check_and_add_debug_xl(volgen_graph_t *graph, dict_t *set_dict, char *volname,
+check_user_xlator_position(dict_t *dict, char *key, data_t *value,
+                           void *prev_xlname)
+{
+    if (strncmp(key, "user.xlator.", strlen("user.xlator.")) != 0) {
```
```suggestion
    if (strncmp(key, "user.xlator.", SLEN("user.xlator.")) != 0) {
```

mykaul commented on xlators/mgmt/glusterd/src/glusterd-volgen.c:
```diff
@@ -2755,6 +2755,114 @@ check_and_add_debug_xl(volgen_graph_t *graph, dict_t *set_dict, char *volname,
+        return false;
+    }
+
+    if (strcmp(data_to_str(value), prev_xlname) == 0) {
```
why do we perform another data_to_str() here, when we've performed it earlier in line 2770?

mykaul commented on xlators/mgmt/glusterd/src/glusterd-volgen.c:
```diff
@@ -2755,6 +2755,114 @@ check_and_add_debug_xl(volgen_graph_t *graph, dict_t *set_dict, char *volname,
+    return false;
+}
+
+int
```
```suggestion
static int
```

mykaul commented on xlators/mgmt/glusterd/src/glusterd-volgen.c:
```diff
@@ -2755,6 +2755,114 @@ check_and_add_debug_xl(volgen_graph_t *graph, dict_t *set_dict, char *volname,
+    return xlator_set_option(xl, optname, strlen(optname), data_to_str(value));
+}
+
+int
```
```suggestion
static int
```

mykaul commented on xlators/mgmt/glusterd/src/glusterd-volgen.c:
```diff
@@ -2755,6 +2755,114 @@ check_and_add_debug_xl(volgen_graph_t *graph, dict_t *set_dict, char *volname,
+{
+    int ret = -1;
+
+    struct {
```
I don't believe we usually define structs like this in Gluster. We usually leave it in .h files or elsewhere, not within a function.

itisravi replied:
They might have seen such a  usage in io_stats_setxattr(). It is okay IMHO and might be an overkill to put it n a header or elsewhere.

xhernandez replied:
The structure is defined and used in two places. I don't think this is good. Given that it's a private structure, it doesn't need to go to a .h file, but we should move the definition outside of the function. At the top of this file would be a good place.

bonfffire replied:
I agree to @xhernandez 's proposal.
So, I'll move the structure to outside of functions.

mykaul commented on xlators/mgmt/glusterd/src/glusterd-volgen.c:
```diff
@@ -2755,6 +2755,114 @@ check_and_add_debug_xl(volgen_graph_t *graph, dict_t *set_dict, char *volname,
+    xlator_t *xl = NULL;
+
+    // convert optkey to xlator type
+    if (gf_asprintf(&type, "user/%s", xlator_name) == -1) {
```
```suggestion
    if (gf_asprintf(&type, "user/%s", xlator_name) < 0) {
```

mykaul commented on xlators/mgmt/glusterd/src/glusterd-volgen.c:
```diff
@@ -2755,6 +2755,114 @@ check_and_add_debug_xl(volgen_graph_t *graph, dict_t *set_dict, char *volname,
+    }
+
+    ret = gf_asprintf(&xlator_option_matcher, "user.xlator.%s.*", xlator_name);
+    if (ret == -1) {
```
```suggestion
    if (ret < 0) {
```

mykaul commented on xlators/mgmt/glusterd/src/glusterd-volgen.c:
```diff
@@ -2755,6 +2755,114 @@ check_and_add_debug_xl(volgen_graph_t *graph, dict_t *set_dict, char *volname,
+    return ret;
+}
+
+int
```
```suggestion
static int
```

mykaul commented on xlators/mgmt/glusterd/src/glusterd-volgen.c:
```diff
@@ -2755,6 +2755,114 @@ check_and_add_debug_xl(volgen_graph_t *graph, dict_t *set_dict, char *volname,
+        goto out;
+    }
+
+    ret = gf_asprintf(&xlator_option_matcher, "user.xlator.%s.*", xlator_name);
```
q: where is xlator_option_matcher freed?

bonfffire replied:
xlator_option_matcher is not freed, so I'll fix it.

itisravi commented on xlators/mgmt/glusterd/src/glusterd-volgen.c:
```diff
@@ -2755,6 +2755,114 @@ check_and_add_debug_xl(volgen_graph_t *graph, dict_t *set_dict, char *volname,
+    }
+
+out:
+    return 0;
```
Is it possible to add some validation if the xlator insertion was unsuccessful? We currently seem to be returning success even if an invalid prev_xlname was given at the `gluster volume set` CLI.

Also, if the insertion was successful, we must display a message at the CLI prompting for a volume restart, just like how it is done for the debug translators.

bonfffire replied:
> Is it possible to add some validation if the xlator insertion was unsuccessful? We currently seem to be returning success even if an invalid prev_xlname was given at the gluster volume set CLI.

Currently, it looks like inserting a debug xlator is not validating.
If set an invalid xlator name to debug.trace, the CLI succeeds and it is recorded in volinfo, but the debug.trace xlator is not inserted in the graph.

```
sh-4.4$ sudo gluster vol set gvol debug.trace unknown-xlator
volume set: success
sh-4.4$ sudo grep 'debug.trace' /var/lib/glusterd/vols/gvol/info
debug.trace=unknown-xlator
```

Should I do a strict validation of input xlname?

If you want to perform validation, we need to have a server xlator list in both CLI and glusterd.
For a better implementation, I think we need to define it in a .h file and share it between cli and glusterd.
And that would be a big fix.

itisravi replied:
That sounds complex. Would something like this work? (I haven't tested it)
```
check_user_xlator_position()
{
    if (set_dict does not contain the key "user.xlator.*")
        return 0;

     if (value == prev_xl_name) {
        insert_user_xlator_to_graph();
        return 0 if successful;
    }
        
     return -EINVAL //value did not match a valid xlator name in the graph.   
        
}
```

bonfffire replied:
I afraid such a code will not work.
set_dict has key-value pairs for all options in volinfo, and server_graph_builder() loops through the hardcoded server xlator list and checks if opt-value matches the xlator name in it.
So if returns an error when opt value does not match, it will not work.

itisravi replied:
Okay, I will go through the current patch set soon.

xhernandez replied:
I would prefer strict checking for user xlator insertion (for debug it would also be great, but this is for another patch). To check the errors, we can use a private structure to pass a return code through dict_foreach functions as the last argument.

bonfffire replied:
I pushed a new commit that changes the volgen process to check if opt-value is invalid, and fail to generate volfile if it is invalid.

itisravi reviewed (changes requested):
Could you also add a test (.t file) please? The test needs to check that the brick volfiles are unchanged when the CLI fails and changes when the CLI succeeds.

itisravi commented:
Triggering regression to see if any of the existing .ts fail.

/run regression

itisravi commented:
/run regression

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-volgen.c:
```diff
@@ -2755,6 +2755,114 @@ check_and_add_debug_xl(volgen_graph_t *graph, dict_t *set_dict, char *volname,
+    if (fnmatch("user.xlator.*.*", key, 0) == 0) {
+        return false;
+    }
+
```
We should make sure that the current `key` has format `user.xlator.*` to avoid problems later in `insert_user_xlator_to_graph`.

itisravi replied:
@xhernandez Do you mean we should allow  even the dot after `user.xlator.$name`? The patch currently seems to allow `user.xlator.posix` but not `user.xlator.posix.one`

xhernandez replied:
If I understand the patch correctly, `user.xlator.<name>` will have the name of the previous xlator after which the user xlator must be loaded. `user.xlator.<name>.<option>` will be used to pass options to the xlator, and this is handled elsewhere. Here we are only searching for the first option, so I think it's correct.

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-volgen.c:
```diff
@@ -2755,6 +2755,114 @@ check_and_add_debug_xl(volgen_graph_t *graph, dict_t *set_dict, char *volname,
+        goto out;
+    }
+
+    gf_log("glusterd", GF_LOG_INFO, "add user xlator=%s to graph", type);
```
Wouldn't `GF_LOG_DEBUG` be better ?

bonfffire replied:
Since modules that do not exist in the Glusterfs source tree will be used, I thought that the INFO level would be better.

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-volgen.c:
```diff
@@ -2755,6 +2755,114 @@ check_and_add_debug_xl(volgen_graph_t *graph, dict_t *set_dict, char *volname,
+    }
+
+    if (strcmp(data_to_str(value), prev_xlname) == 0) {
+        gf_log("glusterd", GF_LOG_INFO,
```
Wouldn't `GF_LOG_DEBUG` be better ?

bonfffire committed:
some fixes based on code review suggestions

- change internal functions to static
- use SLEN() instead of strlen()
- change return value checking for gf_asprintf() to "< 0"

Co-authored-by: Yaniv Kaul <ykaul@redhat.com>

bonfffire committed:
glusterd-volgen: fix memoly leak xlator_option_matcher in insert_user_xlator_to_graph().

bonfffire committed:
glusterd-volgen: move internal structure for dict_foreach_match() in check_and_add_user_xl() to outside of the functions.

bonfffire committed:
glusterd-volgen: remove dead code.

bonfffire committed:
glusterd-volgen: add regression test for user xlator.

This test will verify the following:

- CLI(gluster vol set) will add user/hoge to VOLFILE.
- The user xlator option can be set.
- If a non-existent prev xlname is specified, it will not be added to the volfile.

itisravi commented on tests/basic/user-xlator.t:
```diff
@@ -0,0 +1,42 @@
+TEST $CLI volume set $V0 user.xlator.hoge posix
+TEST grep -q 'user/hoge' ${SERVER_VOLFILE}
+
+TEST $CLI volume start $V0
```
Nit: It might be good to move the volume starting to line 32 and then check that the brick processes are indeed up and running with the custom xlator enabled. Something like `EXPECT_WITHIN $PROCESS_UP_TIMEOUT "1" brick_up_status $V0 $H0 $B0/${V0}1` etc.

itisravi reviewed (commented):
@bonfffire Changes look okay, can  we add the "Please restart the volume for enabling the translator" message to gf_cli_set_volume_cbk() for custom xlator too?

bonfffire committed:
glusterd-volgen: add verification of brick process startup to user xlator tests.

itisravi commented on tests/basic/user-xlator.t:
```diff
@@ -21,18 +21,21 @@ TEST $CLI volume create $V0 replica 3  $H0:$B0/${V0}{1,2,3,4,5,6};
 
+TEST $CLI volume start $V0
+EXPECT_WITHIN $PROCESS_UP_TIMEOUT "1" brick_up_status $V0 $H0 $B0/${V0}1
+
```
It is better to check the status of all bricks.

xhernandez commented on tests/basic/user-xlator.t:
```diff
@@ -0,0 +1,45 @@
+TEST $CLI volume start $V0
+EXPECT_WITHIN $PROCESS_UP_TIMEOUT "1" brick_up_status $V0 $H0 $B0/${V0}1
+
+TEST $CLI volume set $V0 user.xlator.hoge trash
```
After changing the position of the xlator, shouldn't the volume be restarted ?

bonfffire committed:
glusterd: add prompt to restart volume when user xlator is added.

bonfffire committed:
fix clang-format issues.

bonfffire committed:
glusterd-volgen: some improvements to the test code based on code reviews.

Change the following
- Check the status of all bricks
- Make volume restart when user xlator position is changed

bonfffire committed:
glusterd-volgen: strict check prev xlname for user xlator.

bonfffire commented:
/recheck smoke

bonfffire committed:
glusterd-volgen: fix smoke fails of "'for' loop initial declarations are only allowed in C99 mode".

bonfffire committed:
glusterd-volgen: fix smoke fails of "'for' loop initial declarations are only allowed in C99 mode".

bonfffire commented:
/run regression

itisravi commented:
/run regression

itisravi commented on cli/src/cli-rpc-ops.c:
```diff
@@ -1970,6 +1970,60 @@ gf_cli_ganesha_cbk(struct rpc_req *req, struct iovec *iov, int count,
+    int i = 0;
+
+    /*
+     * HACK:
+     *  The server xlator list is also defined here, but it should
+     *  have a common definition with glusterd. However, we are hesitant
+     *  to include the internal header of glusterd.  If want to use the
+     *  shared definition, we need to make a big modification.
```
I was thinking it would be better to send the success/failure message as a key-value pair in the response dict from glusterd, but I realized with your current glusterd side changes, you can rely on `rsp.op_ret` value in  `gf_cli_set_volume_cbk()`. If it is -1, we know its a failure and if its 0, we can form the added_xlator name from the words list itself, without needing this hack or the strcmp.

bonfffire replied:
Thanks for the great idea.
I'll merge the current two separate is_server_debug_xlator and added_user_xlator into added_server_xlator and modify it to only display when the op_ret value is 0.

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-volgen.c:
```diff
@@ -2755,6 +2760,145 @@ check_and_add_debug_xl(volgen_graph_t *graph, dict_t *set_dict, char *volname,
+
+    if (dict_foreach_match(set_dict, check_user_xlator_position, prev_xlname,
+                           insert_user_xlator_to_graph, &data) < 0) {
+        return 1;
```
```suggestion
        return -1;
```

xhernandez commented on cli/src/cli-rpc-ops.c:
```diff
@@ -1970,6 +1970,60 @@ gf_cli_ganesha_cbk(struct rpc_req *req, struct iovec *iov, int count,
+            for (i = 0; i < size; i++) {
+                if (strcmp(value, server_xlators[i]) == 0) {
+                    added_xlator = gf_strdup(key);
+                    break;
```
This exits the `for` loop in line 2013, but not the `while` loop in line 2000. If there's more than one `user.xlator.*` option, we'll only return the last one and we will leak the previous memory assigned to `gf_strdup()`.

xhernandez commented on cli/src/cli-rpc-ops.c:
```diff
@@ -1970,6 +1970,60 @@ gf_cli_ganesha_cbk(struct rpc_req *req, struct iovec *iov, int count,
+        words++;
+    }
+
+    return added_xlator;
```
We don't distinguish between an error and no `user.xlator.*` option found. I think we should handle the error case so that the operation is reported as a failure instead of a success.

bonfffire replied:
I added the error handling and pushed it.
Please review.

bonfffire committed:
glusterd-volgen: change return value to -1 when if failed insert user xlator to graph.

Co-authored-by: Xavi Hernandez <xhernandez@users.noreply.github.com>

bonfffire committed:
glusterd-volgen:  resolve HACK in added_server_xlator()

gf_cli_set_volume_cbk() receives the result of volgen in rsp.op_ret,
which can be used to determine if the server graph has been modified correctly.
Therefore, there is no need to check the value with added_user_xlator or
is_server_debug_xlator, and we should prompt for volume restart only in case of success.

itisravi reviewed (dismissed):
+1

itisravi commented:
/run regression

xhernandez commented on cli/src/cli-rpc-ops.c:
```diff
@@ -1971,48 +1971,41 @@ gf_cli_ganesha_cbk(struct rpc_req *req, struct iovec *iov, int count,
-            if (!(strstr(value, "posix") || strstr(value, "acl") ||
-                  strstr(value, "locks") || strstr(value, "io-threads") ||
-                  strstr(value, "marker") || strstr(value, "index"))) {
+            if (!value || strcmp("client", value) == 0) {
```
If `value` is NULL, we must exit the loop. Otherwise we'll be accessing data outside the expected area.

bonfffire committed:
glusterd-volgen: fix accessing out range of word list in added_server_xlator().

bonfffire committed:
cli: changed to do error handling when it fails to check if the server graph has been changed in "added_server_xlator()".

xhernandez commented on cli/src/cli-rpc-ops.c:
```diff
@@ -1970,49 +1970,119 @@ gf_cli_ganesha_cbk(struct rpc_req *req, struct iovec *iov, int count,
+        }
+
+        if (ret) {
+            *added_xlator = gf_strdup(key);
```
Could you also return -1 if this memory allocation fails ?

xhernandez commented on cli/src/cli-rpc-ops.c:
```diff
@@ -1970,49 +1970,119 @@ gf_cli_ganesha_cbk(struct rpc_req *req, struct iovec *iov, int count,
+        }
+
+        if (ret) {
+            *added_xlator = gf_strdup(key);
```
Same here

bonfffire committed:
cli: make added_server_xlator() returns "-1" if failed to gf_strdup().

xhernandez reviewed (approved):
Thanks for your patience :)

xhernandez commented:
/run regression

xhernandez commented:
@itisravi feel free to merge it if it's also ok to you.
